### PR TITLE
Unified tz solution

### PR DESCRIPTION
**Pull request by shlomi-noach**
Storyline: https://github.com/github/gh-ost/issues/247

This PR identifies and solves the very special case where a `DATETIME` columns is converted to `TIMESTAMP`. Turns out it is a different case from any other `DATETIME`/`TIMESTAMP` related conversion.

The PR fixes handling of this migration cross different timezones; it will behave correctly even as master, replica and app are all on different time zones.

This PR adds a lot of time-related tests.

### CODE REVIEW

File: go/sql/builder.go [hunk 1]
```diff
@@ -83,13 +97,19 @@ func BuildEqualsPreparedComparison(columns []string) (result string, err error)
...
+	if columns.Len() == 0 {
 		return "", fmt.Errorf("Got 0 columns in BuildSetPreparedClause")
 	}
 	setTokens := []string{}
-	for _, column := range columns {
-		setTokens = append(setTokens, fmt.Sprintf("%s=?", EscapeName(column)))
+	for _, column := range columns.Columns() {
+		var setToken string
+		if column.timezoneConversion != nil {
+			setToken = fmt.Sprintf("%s=convert_tz(?, '%s', '%s')", EscapeName(column.Name), column.timezoneConversion.ToTimezone, "+00:00")
```
Comment (jonahberquist) [inline go/sql/builder.go]: Nice, glad this works 👍

**Comment by shlomi-noach** [comment]
Add a (enforced, tested) limitation that a column converting from `DATETIME` to `TIMESTAMP` cannot be part of the shared unique key `gh-ost` iterates on.
This limitation can be lifted in the future given more development.

**Comment by shlomi-noach** [comment]
Sent to production testing

**Comment by shlomi-noach** [comment]
Tests are good. Merging.

**Pull request merged**